lucene/core/src/java/org/apache/lucene/search/VectorDistanceQuery.java

```diff
@@ -0,0 +1,185 @@
+  }
+
+  private boolean equalsTo(VectorDistanceQuery other) {
+    return Objects.equals(field, other.field) &&
+        Arrays.equals(queryVector, other.queryVector) &&
+        Objects.equals(numProbes, other.numProbes);
+  }
+
+  private static class VectorDistanceWeight extends Weight {
+    private final ScoreMode scoreMode;
+    private final float boost;
+
+    private final String field;
+    private final float[] queryVector;
+    private final int numProbes;
+
+    VectorDistanceWeight(Query query, float boost, ScoreMode scoreMode,
+                         String field, float[] queryVector, int numProbes) {
+      super(query);
+      this.scoreMode = scoreMode;
+      this.boost = boost;
+
+      this.field = field;
+      this.queryVector = queryVector;
+      this.numProbes = numProbes;
+    }
+
+    @Override
+    public Explanation explain(LeafReaderContext context, int doc) throws IOException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Scorer scorer(LeafReaderContext context) throws IOException {
+      VectorValues vectorValues = context.reader().getVectorValues(field);
+      TermsEnum clusters = vectorValues.getClusterPostings().iterator();
+      BinaryDocValues docValues = vectorValues.getVectorValues();
+
+      List<BytesRef> closestCentroids = findClosestCentroids(clusters);
+
+      List<Scorer> subScorers = new ArrayList<>();
+      for (BytesRef encodedCentroid : closestCentroids) {
+        boolean seekSuccess = clusters.seekExact(encodedCentroid);
+        assert seekSuccess;
+
+        Similarity.SimScorer simScorer = new BooleanSimilarity().scorer(1.0f, null);
+        LeafSimScorer leafSimScorer = new LeafSimScorer(simScorer, context.reader(), field, false);
+        TermScorer termScorer = new TermScorer(this, clusters.postings(null), leafSimScorer);
+        subScorers.add(termScorer);
+      }
+
+      return new DisjunctionScorer(this, subScorers, scoreMode) {
+        @Override
+        protected float score(DisiWrapper topList) throws IOException {
+          int docId = docValues.advance(topList.doc);
+          assert docId == topList.doc;
+
+          BytesRef encodedVector = docValues.binaryValue();
+          double dist = VectorField.l2norm(queryVector, encodedVector);
+          return (float) (boost / (1.0 + dist));
+        }
+
+        @Override
+        public float getMaxScore(int upTo) {
+          return boost;
+        }
+      };
+    }
+
+    private List<BytesRef> findClosestCentroids(TermsEnum centroids) throws IOException {
+      PriorityQueue<Map.Entry<BytesRef, Double>> queue = new PriorityQueue<>(
+          (first, second) -> -1 * Double.compare(first.getValue(), second.getValue()));
+
+      while (true) {
+        BytesRef encodedCentroid = centroids.next();
+        if (encodedCentroid == null) {
+          break;
+        }
+
+        double dist = VectorField.l2norm(queryVector, encodedCentroid);
```

Comment:
do we also consider to have a function `l2normsquared` that can be used for cases where we don't care about scores, just ranks. For example for this case.